Incorrect bounds check on ptrCast-ed const value

```zig
const std = @import("std");
const minInt = std.math.minInt;

test "aoeu" {
    const is_value = ~@intCast(isize, minInt(isize));
    const is_bytes = @ptrCast([*]const u8, &is_value)[0..@sizeOf(isize)];
}
```
Produces:
```
error: out of bounds slice
```
The compiler seems to think the bounds are `[0..1]`. Changing `is_value` to var makes the code compile correctly.

Looks like a user on reddit ran into this error: https://www.reddit.com/r/Zig/comments/ju0tqw/out_of_bounds_error_when_taking_a_slice_of_a/

Here is the simplified example I came up with:
```zig
test "" {
    const val : u32 = 0;
    const val_ptr = @ptrCast([*]const u8, &val);
    _ = val_ptr[1];
}
```
```
./test.zig:4:16: error: index 1 outside pointer of size 1
    _ = val_ptr[1];
```



This example has a better error message that confirms what @alexnask said:
> What is happening is that the casted pointer still points to the correct spot in memory but the information on the underlying byte count is lost. That means that the compiler thinks this is a pointer to a single byte.

--- a/test/behavior/error.zig
+++ b/test/behavior/error.zig
@@ -287,6 +287,27 @@ test "inferred empty error set comptime catch" {
     S.foo() catch @compileError("fail");
 }
 
+test "error inference with an empty set" {
+    const S = struct {
+        const Struct = struct {
+            pub fn func() (error{})!usize {
+                return 0;
+            }
+        };
+
+        fn AnotherStruct(comptime SubStruct: type) type {
+            return struct {
+                fn anotherFunc() !void {
+                    try expect(0 == (try SubStruct.func()));
+                }
+            };
+        }
+    };
+
+    const GeneratedStruct = S.AnotherStruct(S.Struct);
+    try GeneratedStruct.anotherFunc();
+}
+
 test "error union peer type resolution" {
     if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest; // TODO
     if (builtin.zig_backend == .stage2_sparc64) return error.SkipZigTest; // TODO

--- a/test/behavior/ptrcast.zig
+++ b/test/behavior/ptrcast.zig
@@ -203,6 +203,37 @@ test "comptime ptrcast keeps larger alignment" {
     }
 }
 
+test "ptrcast of const integer has the correct object size" {
+    const is_value = ~@intCast(isize, std.math.minInt(isize));
+    const is_bytes = @ptrCast([*]const u8, &is_value)[0..@sizeOf(isize)];
+    if (@sizeOf(isize) == 8) {
+        switch (native_endian) {
+            .Little => {
+                try expect(is_bytes[0] == 0xff);
+                try expect(is_bytes[1] == 0xff);
+                try expect(is_bytes[2] == 0xff);
+                try expect(is_bytes[3] == 0xff);
+
+                try expect(is_bytes[4] == 0xff);
+                try expect(is_bytes[5] == 0xff);
+                try expect(is_bytes[6] == 0xff);
+                try expect(is_bytes[7] == 0x7f);
+            },
+            .Big => {
+                try expect(is_bytes[0] == 0x7f);
+                try expect(is_bytes[1] == 0xff);
+                try expect(is_bytes[2] == 0xff);
+                try expect(is_bytes[3] == 0xff);
+
+                try expect(is_bytes[4] == 0xff);
+                try expect(is_bytes[5] == 0xff);
+                try expect(is_bytes[6] == 0xff);
+                try expect(is_bytes[7] == 0xff);
+            },
+        }
+    }
+}
+
 test "implicit optional pointer to optional anyopaque pointer" {
     if (builtin.zig_backend == .stage2_arm) return error.SkipZigTest; // TODO
     if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest; // TODO

